File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    int findLeft(vector<int>& nums, int target){
        int left = 0, right = nums.size() - 1, ans = -1;
        while(left <= right){
            int mid = left + (right - left) / 2;
            if(nums[mid] == target){
                ans = mid;
                right = mid - 1;
            }
            else if(nums[mid] > target) right = mid - 1;
            else left = mid + 1;
        }
        return ans;
    }

    int findRight(vector<int>& nums, int target){
        int left = 0, right = nums.size() - 1, ans = -1;
        while(left <= right){
            int mid = left + (right - left) / 2;
            if(nums[mid] == target){
                ans = mid;
                left = mid + 1;
            }
            else if(nums[mid] > target) right = mid - 1;
            else left = mid + 1;
        }
        return ans;
    }

    vector<int> searchRange(vector<int>& nums, int target) {
        return {findLeft(nums, target), findRight(nums, target)};
    }
};

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
class Solution(object):
    def search_left(self, nums, target):
        l, r = 0, len(nums) - 1
        pos = -1
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] == target:
                pos = mid
                r = mid - 1   
            elif nums[mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        return pos

    def search_right(self, nums, target):
        l, r = 0, len(nums) - 1
        pos = -1
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] == target:
                pos = mid
                l = mid + 1   # keep searching right
            elif nums[mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        return pos

    def searchRange(self, nums, target):
        return [self.search_left(nums, target), self.search_right(nums, target)]
